app/scripts/controllers/transactions/index.js

```diff
@@ -356,42 +356,44 @@ export default class TransactionController extends EventEmitter {
       opts.id,
     );
 
-    // listen for tx completion (success, fail)
-    return new Promise((resolve, reject) => {
-      this.txStateManager.once(
-        `${initialTxMeta.id}:finished`,
-        (finishedTxMeta) => {
-          switch (finishedTxMeta.status) {
-            case TransactionStatus.submitted:
-              return resolve(finishedTxMeta.hash);
-            case TransactionStatus.rejected:
-              return reject(
-                cleanErrorStack(
-                  ethErrors.provider.userRejectedRequest(
-                    'MetaMask Tx Signature: User denied transaction signature.',
-                  ),
-                ),
-              );
-            case TransactionStatus.failed:
-              return reject(
-                cleanErrorStack(
-                  ethErrors.rpc.internal(finishedTxMeta.err.message),
-                ),
-              );
-            default:
-              return reject(
-                cleanErrorStack(
-                  ethErrors.rpc.internal(
-                    `MetaMask Tx Signature: Unknown problem: ${JSON.stringify(
-                      finishedTxMeta.txParams,
-                    )}`,
-                  ),
-                ),
-              );
-          }
-        },
-      );
-    });
+    const txId = initialTxMeta.id;
+    const isCompleted = initialTxMeta.status !== TransactionStatus.unapproved;
+
+    const finishedPromise = isCompleted
+      ? Promise.resolve(initialTxMeta)
+      : this._waitForTransactionFinished(txId);
+
+    if (!isExisting && !isCompleted) {
+      try {
+        await this._requestTransactionApproval(initialTxMeta);
+      } catch (error) {
+        // Errors generated from final status using finished event
+      }
```

Comment:
This is a farfetched case, but if the `_requestTransactionApproval` ends up throwing after setting the transaction status to approved but before setting it to submitted, then we will end up with a pending Promise (`finishedPromise`) that will never resolve and will be blocking the execution flow. Should we or even do we want to handle transactions that were not submitted but were a approved? 🤔 

(I guess that this could already happen before anyway, so any solution should be done on a subsequent PR)

Reply:
Good spot, I've fixed the condition inside the error handling in `_requestTransactionApproval` so it will now correctly cancel or fail the transaction if it's in any incomplete status. Meaning any errors here are just for information, and the transaction will always be `submitted`, `rejected`, or `failed` at this point.